### PR TITLE
swig: update version 3.0.12->4.0.0

### PR DESCRIPTION
**Pull request by MarcusCalhoun-Lopez**
###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.14.4
Xcode 10.2.1 

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by yan12125** [comment]
Sorry, I didn't notice this PR before hitting the merge button for my previous PR ("swig: drop php 5.2 variant"). There's now a conflict.

**Comment by michaelld** [comment]
How thoroughly have you tested the new SWIG4 to make sure it is compatible with existing projects?

**Comment by michaelld** [comment]
Given how long it's been since the last release & the number of changes in both bug fixes and code additions, I'm concerned about compatibility with existing ports. My preference here is a separate SWIG4 port that's installable alongside SWIG (3), and one then can `select` which one to use [I almost have this working, BTW]. By having both installable SWIGs, we can select between them for testing / evaluation purposes & work toward updating / fixing ports from 3 to 4. Once that's done, SWIG (3) could be removed; and in the interim we could (somehow) mark SWIG (3) as deprecated in favo(u)r of SWIG4.

**Comment by yan12125** [comment]
FWIW, there is a bug report for Arch Linux that subversion 1.12.0 does not build with swig 4.0.0: https://bugs.archlinux.org/task/62513.

**Comment by MarcusCalhoun-Lopez** [comment]
> How thoroughly have you tested the new SWIG4 to make sure it is compatible with existing projects?

I mostly tested that the new support for Octave 5.1 was working.
Some ports and projects that have been broken for a while are now working, so I was hoping we could upgrade SWIG sooner rather than later.
Of course, you are correct that breaking many ports for the sake a small number is no win.

**Comment by MarcusCalhoun-Lopez** [comment]
> My preference here is a separate SWIG4 port that's installable alongside SWIG (3), and one then can `select` which one to use [I almost have this working, BTW].

Should I close this pull request in favor of what you are working on?

**Comment by michaelld** [comment]
OK; good to know about SVN and Octave. Given that there are some 180 ports that depend on SWIG, I think it's wisest to move to a WIP PR for what I'm trying to do. We can test there & fix some, leave others on SWIG3 that just don't work with SWIG4 & aren't easily upgradeable. I think the challenge will be fixing ports to use the correct SWIG binary, since we can't rely on the one "select"ed. Hence why I think this is a WIP PR for a bit. Can I somehow force push over this PR's commits? We could then just leave this PR as is & rename it & work on it here. If not, then I can always open a new PR & tag it here.

**Comment by MarcusCalhoun-Lopez** [comment]
> Can I somehow force push over this PR's commits? We could then just leave this PR as is & rename it & work on it here. If not, then I can always open a new PR & tag it here.

When I opened this pull request, I checked the box "Allow edits from maintainers."
However, I have never used this feature.

However you want to proceed is fine with me.

**Comment by michaelld** [comment]
> When I opened this pull request, I checked the box "Allow edits from maintainers."
> However, I have never used this feature. However you want to proceed is fine with me.

OK I'll see if I can force push over yours here. If that's not obvious enough then I'll just open a new PR.

**Comment by kencu** [comment]
To force push over another persons PR, the only method that has ever worked for me is to clone their personal macports-ports repo locally temporarily, and work with that just like I would work with my own. That works easily for me, and I can force-push to their PR branch.

A number of people have told me about ways that are supposed to work without cloning their entire repo, but I can never get any of those other ways to work when I try them. This does work, and it only takes a minute or two to clone their entire repo, so ... that's what I do.

**Comment by michaelld** [comment]
@kencu that's exactly what I thought would work & I'm making it so ... just not enough time in the day to get all the things done I'd like! This one is a little easier than the Boost one, so I'll probably do it first. But the Boost one will get stale pretty quickly ... trade-offs!

**Comment by luzpaz** [comment]
any progress on this ?

**Comment by michaelld** [comment]
Not much. My work queue overfloweth ... not enough of me for the amount of time allocated! I'd welcome a PR if anyone has time ...

**Comment by MarcusCalhoun-Lopez** [comment]
> Not much. My work queue overfloweth ... not enough of me for the amount of time allocated! I'd welcome a PR if anyone has time ...

Please correct me if I am mistaken, but the current PR is insufficient.
Correct?
You are looking for a PR that allows both version of swig to coexist?

**Comment by michaelld** [comment]
I'd prefer to keep SWIG 3 around until it's clear that it's not needed. I'll admit I haven't tried building swig-dependent ports using the new SWIG 4 ... maybe, like Boost, they will pretty much work & if that's the case then I'd be fine with just using this PR. I'm just wanting to be cautious because SWIG is so critical to some projects.

**Comment by MarcusCalhoun-Lopez** [comment]
How is this for an upgrade path?
- Create a new Portfile called swig3 (swig version 3).
- Update swig to version 4 but rename `swig` and `ccache-swig` to `swig4` and `ccache-swig4` to avoid conflict with swig3.
- Have swigXYZ depends on swig3XYZ.
    - No dependencies need to be changed.
    - By default, dependencies will keep using SWIG 3.
- Ports that need SWIG 4 will have to convinced to use `${prefix}/bin/swig4`.
- As soon as we are confident that all ports work with SWIG 3, we can simply remove it.

**Comment by MarcusCalhoun-Lopez** [comment]
@michaelld, do you have any thoughts about this latest pull request?
Do you have any objections about merging it?

**Comment by michaelld** [comment]
@MarcusCalhoun-Lopez Sorry for the delay. I think your proposed upgrade path is sufficient. I'd prefer to not just plow ahead with merging this PR given how little we know about how SWIG4 will perform compared with SWIG3, since we have no "devel" version for incremental testing (as we do for some other critical ports such as CMake; swig-devel just is not practical without a PG and maybe some other tweaks). Keeping current ports using SWIG3 is IMHO wise. We can then coerce those ports that can to move to SWIG4 as time allows ... and then deprecating and eventually removing SWIG3 in favor of SWIG4 a little down the road to give projects time to update.

**Comment by michaelld** [comment]
We can even leave current SWIG alone, and just create SWIG4 related ports that install separately from SWIG(3).

**Comment by MarcusCalhoun-Lopez** [comment]
@michaelld, thank you for your response.
I am afraid I was not clear in my earlier description.

The current pull request keeps **all** ports using Swig 3.

The only way a port will use Swig 4 is if it specifically requests it
(e.g. `-DSWIG_EXECUTABLE=${prefix}/bin/swig4` or `configure.env-append SWIG=${prefix}/bin/swig4`).

If I have set this up correctly (always a big if), there should be no disruption to any ports that use Swig.
I am sorry if I have not explained the changes well.
I would be happy to go into more detail if needed.

**Comment by michaelld** [comment]
Let me do another review of this PR then. I thought it was just moving SWIG to SWIG4 ... guessing I need to pay attention here a little more!

**Comment by michaelld** [comment]
Ah OK ... clever & I like it! Let me give this PR a testing locally to verify that what you write & I now think it should be doing is what it's doing. My bad for not paying attention closely here ... sorry about that!

File: devel/swig3/Portfile [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    most commonly used to create high-level interpreted programming \
+    environments, user interfaces, and as a tool for testing and \
+    prototyping C/C++ software. SWIG may be freely used, distributed, and \
+    modified for commercial and noncommercial use.
+
+homepage        http://www.swig.org/
+platforms       darwin
+master_sites    sourceforge:project/swig/swig/swig-${version}
+
+distfiles       swig-${version}${extract.suffix}
```
Comment (michaelld) [inline devel/swig3/Portfile]: Can you add in ```dist_subdir     swig``` here? That way there is continuity to where SWIG source code tarballs are stored?

**Comment by MarcusCalhoun-Lopez** [comment]
Thank you for the evaluation.

**Pull request merged**